pandas/core/dtypes/concat.py

```diff
@@ -108,7 +173,17 @@ def is_nonempty(x) -> bool:
         # we ignore axis here, as internally concatting with EAs is always
         # for axis=0
         if not single_dtype:
-            target_dtype = find_common_type([x.dtype for x in to_concat])
+            # Special case for handling concat with categorical series.
+            # We need to make sure that categorical dtype is preserved
+            # when an array of valid values is given (GH#25383)
+            if (
+                isinstance(to_concat[0], ExtensionArray)
+                and all(x.shape[0] == 1 for x in to_concat[1:])
```

Comment:
I couldn't find a better way to detect when the `concat_compat` function is called through index expansion, so in cases like this:
```python
ser = pd.Series(Categorical(["a", "b", "c"]))
ser.loc[3] = "c"
```
With the latest commit we are raising a `ValueError` when an invalid value is added to the categorical through index expansion. it also enables the index expansion of a categorical of any `dtype`.